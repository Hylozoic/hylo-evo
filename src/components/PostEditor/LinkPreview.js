import React, { PropTypes } from 'react'

export default function LinkPreview ({className, linkPreview: { title, description, imageUrl, id }}) {
  return <div className={className}>
    {title}
    <img src={imageUrl} />
    {description}
  </div>
}
LinkPreview.propTypes = {
  linkPreview: PropTypes.object.isRequired,
  className: PropTypes.string
}
