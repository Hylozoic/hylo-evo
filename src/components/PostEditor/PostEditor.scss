// mixins

.section-label {
  color: $color-rhino-80;
  white-space: nowrap;
  margin-right: 28px;
  font-size: 14px;
}

// Overall wrapper

.wrapper {
  display: flex;
  flex-direction: column;
  min-width: $post-editor-width;
  list-style: none;
  border: solid 1px rgb(0,0,0);
  border-radius: $space-1x;
  border: 1px solid $color-cape-cod-10;
  background: $color-white;
}

// Header area

.header {
  padding: $space-4x $space-4x 0 $space-4x;
}

.initial {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  line-height: normal;
  margin-bottom: $space-4x;
  &-prompt {
    composes: bdy-lt-sm from 'css/typography.scss';
  }
  &-closeButton {
    cursor: pointer;
  }
}

.postTypes {
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: $space-6x;
}

.postType {
  composes: tag from 'css/typography.scss';
  font-size: 11px;
  letter-spacing: 1.2px;
  line-height: 40px;
  display: inline-block;
  width: 100%;
  margin: 0 $space-2x;
  padding: 0 $space-5x;
  text-align: center;
  border-radius: $space-1x;
  background-color: $color-athens-gray-dark;
  &:last-child {
    margin-right: 0;
  }
  &:first-child {
    margin-left: 0;
  }
  &-discussion {
    &.selectable:hover, &:active, &.active {
      color: $color-picton-blue;
      background-color: $color-link-water;
    }
  }
  &-request {
    &.selectable:hover, &:active, &.active {
      color: $color-persimmon;
      background-color: $color-peach-schnapps;
    }
  }
  &-offer {
    &.selectable:hover, &:active, &.active  {
      color: $color-caribbean-green;
      background-color: $color-iceberg;
    }
  }
}

// Body area

.body {
  padding: 0 $space-4x;
  display: flex;
  flex-direction: row;
  &-column:last-child {
    width: 100%;
  }
}

.titleAvatar {
  margin-right: $space-2x;
  padding: $space-1x;
  height: $space-9x;
}

.titleInput {
  composes: hdr-headline from 'css/typography.scss';
  margin-bottom: $space-2x;
  padding: 0;
  height: $space-9x;
  width: 100%;
  outline: none;
  border: none;
  &:disabled {
    color: $color-rhino-80;
  }
  &::placeholder {
    color: $color-rhino-30;
  }
}

.editor {
  height: 200px;
  margin-bottom: $space-6x;
  color: $color-rhino-80;
}

.linkPreview {
  height: $space-16x;
  background-color: $color-medium-purple;
}

// Footer area

.footer {
  padding: 0;
}

.postIn {
  min-height: 3rem;
  border-top: solid 1px $color-ghost-30;
  padding: $space-5x;
  line-height: 1;
  display: flex;
  align-items: center;
  &-label {
    composes: section-label;
  }
  &-communities {
    /* !empty */
  }
}

.actionsBar {
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
  padding: $space-3x 18px;
  align-items: center;
  border-top: solid 1px $color-ghost-30;
}

.actions {
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
}

.action-icon {
  font-size: 32px;
  cursor: pointer;
  &:hover {
    color: $color-caribbean-green;
  }
}

.highlight-icon {
  color: $color-caribbean-green;
  cursor: default;
}

.postButton {
  margin-left: auto;
}
