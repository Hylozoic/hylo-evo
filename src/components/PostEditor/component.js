import React from 'react'
import './component.scss'
import PostLabel from 'components/PostLabel'
import Avatar from 'components/Avatar'
import HyloEditor from 'components/HyloEditor'
import Button from 'components/Button'

const TITLE_PLACEHOLDER = 'Whatâ€™s on your mind?'
const BODY_PLACEHOLDER = 'Add a description'

export default class PostEditor extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      title: ''
    }
  }

  handleTitleChange = (event) => this.setState({title: event.target.value})

  render () {
    const { title } = this.state

    return <div styleName='wrapper'>
      <div styleName='body'>
        <div styleName='initialPrompt' className='bdy-lt-sm'>What are you looking to post?</div>
        <div styleName='postTypes'>
          <PostLabel type='discussion' styleName='postType postType-left' />
          <PostLabel type='request' styleName='postType' />
          <PostLabel type='offer' styleName='postType postType-right' />
        </div>
        <div styleName='title'>
          <Avatar medium
            styleName='title-avatar'
            url=''
            avatarUrl='https://d3ngex8q79bk55.cloudfront.net/user/13986/avatar/1444260480878_AxolotlPic.png' />
          <input type='text'
            styleName='title-input'
            placeholder={TITLE_PLACEHOLDER}
            value={title}
            onChange={this.handleTitleChange} />
        </div>
        <HyloEditor styleName='editor' debug placeholder={BODY_PLACEHOLDER} />
      </div>
      <div styleName='footer'>
        <div styleName='postIn'>
          <div styleName='postIn-label'>
            Post in
          </div>
          <div styleName='postIn-communities'>
            ...
          </div>
        </div>
        <div styleName='actionsBar'>
          <div styleName='actions'>
            ...
          </div>
          <Button styleName='postButton' label='Post' color='green' />
        </div>
      </div>
    </div>
  }
}
