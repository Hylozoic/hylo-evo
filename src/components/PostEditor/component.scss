.wrapper {
  display: flex;
  flex-direction: column;
  width: 590px;
  list-style: none;
  border: solid 1px rgb(0,0,0);
  border-radius: $space-1x;
  border: 1px solid $color-cape-cod-10;
}

.body {
  padding: $space-4x;
}

.initialPrompt {
  margin-bottom: $space-4x;
}

.postTypes {
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
  justify-content: space-between;
  margin-bottom: $space-6x;
}

.postType {
  width: 100%;
  font-size: 0.625rem;
  line-height: 0;
  margin: 0 $space-2x;
  padding: $space-5x;
  border-radius: $space-1x;
  &-right {
    margin-right: 0;
  }
  &-left {
    margin-left: 0;
  }
}

.title {
  border: none;
  margin-bottom: $space-2x;
  // TODO: Wrap the title input and HyloEditor and flexbox this layout
  margin-left: $space-12x;
  &-avatar {
    margin-left: -$space-12x;
    margin-right: $space-2x;
  }
  &-input {
    composes: hdr-headline from 'typography.scss';
    outline: none;
    border: none;
  }
}

.editor {
  height: 200px;
  margin-left: $space-12x;
  margin-bottom: $space-6x;
}

// Footer area

.footer {
  /* !empty */
}

.postIn {
  height: 3rem;
  border-top: solid 1px $color-ghost-30;
  padding: $space-3x;
  &-label {
    display: inline-block;
    margin-right: $space-2x;
  }
  &-communities {
    display: inline-block;
  }
}

.actionsBar {
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
  justify-content: space-between;
  height: 3.75rem;
  padding: $space-3x;
  background-color: $color-athens-gray-dark;
}

.actions {
  /* !empty */
}

.postButton {
  /* !empty */
}
