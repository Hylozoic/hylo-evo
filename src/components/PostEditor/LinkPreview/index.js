import LinkPreview from './LinkPreview'
export default LinkPreview
