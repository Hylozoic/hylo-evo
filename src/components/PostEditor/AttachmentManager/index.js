import component from './AttachmentManager'
import connector from './AttachmentManager.connector'

export default connector(component)
