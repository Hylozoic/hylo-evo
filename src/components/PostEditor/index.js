import component from './PostEditor'
import connector from './PostEditor.connector'

export default connector(component)
