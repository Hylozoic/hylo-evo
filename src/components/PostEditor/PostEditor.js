import PropTypes from 'prop-types'
import React from 'react'
import ReactTooltip from 'react-tooltip'
import { get, isEqual, throttle } from 'lodash/fp'
import cx from 'classnames'
import styles from './PostEditor.scss'
import contentStateToHTML from 'components/HyloEditor/contentStateToHTML'
import Icon from 'components/Icon'
import RoundImage from 'components/RoundImage'
import HyloEditor from 'components/HyloEditor'
import Button from 'components/Button'
import CommunitiesSelector from 'components/CommunitiesSelector'
import TopicSelector from 'components/TopicSelector'
import LinkPreview from './LinkPreview'
import ChangeImageButton from 'components/ChangeImageButton'
import SendAnnouncementModal from 'components/SendAnnouncementModal'
import AttachmentManager from './AttachmentManager'
import { uploadSettings } from './AttachmentManager/AttachmentManager'
import cheerio from 'cheerio'
import { TOPIC_ENTITY_TYPE } from 'hylo-utils/constants'
import { MAX_TITLE_LENGTH } from './PostEditor.store'

export default class PostEditor extends React.Component {
  static propTypes = {
    initialPrompt: PropTypes.string,
    onClose: PropTypes.func,
    titlePlaceholderForPostType: PropTypes.object,
    detailsPlaceholder: PropTypes.string,
    communityOptions: PropTypes.array,
    currentUser: PropTypes.object,
    currentCommunity: PropTypes.object,
    post: PropTypes.shape({
      id: PropTypes.string,
      type: PropTypes.string,
      title: PropTypes.string,
      details: PropTypes.string,
      linkPreview: PropTypes.object,
      communities: PropTypes.array
    }),
    linkPreviewStatus: PropTypes.string,
    createPost: PropTypes.func,
    updatePost: PropTypes.func,
    pollingFetchLinkPreview: PropTypes.func,
    removeLinkPreview: PropTypes.func,
    clearLinkPreview: PropTypes.func,
    goToPost: PropTypes.func,
    editing: PropTypes.bool,
    loading: PropTypes.bool
  }

  static defaultProps = {
    initialPrompt: 'What are you looking to post?',
    titlePlaceholderForPostType: {
      offer: 'What super powers can you offer?',
      request: 'What are you looking for help with?',
      default: 'Whatâ€™s on your mind?'
    },
    detailsPlaceholder: 'Add a description',
    post: {
      type: 'discussion',
      title: '',
      details: '',
      communities: []
    },
    editing: false,
    loading: false
  }

  buildStateFromProps = ({ editing, loading, currentCommunity, post, topic, announcementSelected }) => {
    const defaultPostWithCommunitiesAndTopic = Object.assign({}, PostEditor.defaultProps.post, {
      communities: currentCommunity ? [currentCommunity] : [],
      topics: topic ? [topic] : [],
      detailsTopics: []
    })

    const currentPost = post || defaultPostWithCommunitiesAndTopic
    return {
      post: currentPost,
      titlePlaceholder: this.titlePlaceholderForPostType(currentPost.type),
      valid: editing === true, // if we're editing, than it's already valid upon entry.
      announcementSelected: announcementSelected,
      toggleAnnouncementModal: false,
      titleLengthError: false
    }
  }

  constructor (props) {
    super(props)
    this.state = this.buildStateFromProps(props)
  }

  componentDidMount () {
    setTimeout(() => {
      this.titleInput.focus()
    }, 100)
  }

  componentDidUpdate (prevProps) {
    const linkPreview = this.props.linkPreview
    if (get('post.id', this.props) !== get('post.id', prevProps)) {
      this.reset(this.props)
      this.editor.focus()
    } else if (linkPreview !== prevProps.linkPreview) {
      this.setState({
        post: {...this.state.post, linkPreview}
      })
    }
  }

  componentWillUnmount () {
    this.props.clearLinkPreview()
  }

  reset = (props) => {
    this.editor.reset()
    this.communitiesSelector.reset()
    this.setState(this.buildStateFromProps(props))
  }

  focus = () => this.editor && this.editor.focus()

  handlePostTypeSelection = event => {
    const type = event.target.textContent.toLowerCase()
    this.setState({
      post: {...this.state.post, type},
      titlePlaceholder: this.titlePlaceholderForPostType(type),
      valid: this.isValid({ type })
    })
  }

  titlePlaceholderForPostType (type) {
    const { titlePlaceholderForPostType } = this.props
    return titlePlaceholderForPostType[type] || titlePlaceholderForPostType['default']
  }

  postTypeButtonProps = (forPostType) => {
    const { loading } = this.props
    const { type } = this.state.post
    const active = type === forPostType
    const className = cx(
      styles['postType'],
      styles[`postType-${forPostType}`],
      {
        [styles[`active`]]: active,
        [styles[`selectable`]]: !loading && !active
      }
    )
    return {
      label: forPostType,
      onClick: this.handlePostTypeSelection,
      disabled: loading,
      color: '',
      className
    }
  }

  handleTitleChange = (event) => {
    const title = event.target.value
    title.length >= MAX_TITLE_LENGTH ? this.setState({titleLengthError: true}) : this.setState({titleLengthError: false})
    this.setState({
      post: {...this.state.post, title},
      valid: this.isValid({ title })
    })
  }

  handleDetailsChange = (editorState, contentChanged) => {
    if (contentChanged) {
      const contentState = editorState.getCurrentContent()
      this.setLinkPreview(contentState)
      this.updateTopics(contentState)
    }
  }

  setLinkPreview = (contentState) => {
    const { pollingFetchLinkPreview, linkPreviewStatus, clearLinkPreview } = this.props
    const { linkPreview } = this.state.post
    if (!contentState.hasText() && linkPreviewStatus) return clearLinkPreview()
    if (linkPreviewStatus === 'invalid' || linkPreviewStatus === 'removed') return
    if (linkPreview) return
    pollingFetchLinkPreview(contentStateToHTML(contentState))
  }

  updateTopics = throttle(2000, (contentState) => {
    const html = contentStateToHTML(contentState)
    const $ = cheerio.load(html)
    var topicNames = []
    $(`a[data-entity-type=${TOPIC_ENTITY_TYPE}]`).map((i, el) =>
      topicNames.push($(el).text().replace('#', '')))
    const hasChanged = !isEqual(this.state.detailsTopics, topicNames)
    if (hasChanged) {
      this.setState({
        detailsTopics: topicNames.map(tn => ({name: tn, id: tn}))
      })
    }
  })

  removeLinkPreview = () => {
    this.props.removeLinkPreview()
    this.setState({
      post: {...this.state.post, linkPreview: null}
    })
  }

  setSelectedCommunities = communities => {
    this.setState({
      post: {...this.state.post, communities},
      valid: this.isValid({ communities })
    })
  }

  isValid = (postUpdates = {}) => {
    const { type, title, communities } = Object.assign({}, this.state.post, postUpdates)
    return !!(this.editor &&
      communities &&
      type.length > 0 &&
      title.length > 0 &&
      communities.length > 0 &&
      title.length <= MAX_TITLE_LENGTH
    )
  }

  save = () => {
    const { editing, createPost, updatePost, onClose, goToPost, images, files, setAnnouncement, announcementSelected } = this.props
    const { id, type, title, communities, linkPreview } = this.state.post
    const details = this.editor.getContentHTML()
    const topicNames = this.topicSelector.getSelected().map(t => t.name)
    const postToSave = {
      id, type, title, details, communities, linkPreview, imageUrls: images, fileUrls: files, topicNames, sendAnnouncement: announcementSelected
    }
    const saveFunc = editing ? updatePost : createPost
    setAnnouncement(false)
    saveFunc(postToSave).then(editing ? onClose : goToPost)
  }

  buttonLabel = () => {
    const { postPending, editing } = this.props
    if (postPending) return 'Posting...'
    if (editing) return 'Save'
    return 'Post'
  }

  toggleAnnouncementModal = () => {
    const { showAnnouncementModal } = this.state
    this.setState({
      ...this.state,
      showAnnouncementModal: !showAnnouncementModal
    })
  }

  render () {
    const { titlePlaceholder, titleLengthError, valid, post, detailsTopics = [], showAnnouncementModal } = this.state
    const { id, title, details, communities, linkPreview, topics } = post
    const {
      onClose, initialPrompt, detailsPlaceholder,
      currentUser, communityOptions, loading, addImage,
      showImages, addFile, showFiles, setAnnouncement, announcementSelected,
      canModerate, myModeratedCommunities
    } = this.props
    return <div styleName={showAnnouncementModal ? 'hide' : 'wrapper'} ref={element => { this.wrapper = element }}>
      <div styleName='header'>
        <div styleName='initial'>
          <div styleName='initial-prompt'>{initialPrompt}</div>
          <a styleName='initial-closeButton' onClick={onClose}><Icon name='Ex' /></a>
        </div>
        <div styleName='postTypes'>
          <Button {...this.postTypeButtonProps('discussion')} />
          <Button {...this.postTypeButtonProps('request')} />
          <Button {...this.postTypeButtonProps('offer')} />
        </div>
      </div>
      <div styleName='body'>
        <div styleName='body-column'>
          <RoundImage
            medium
            styleName='titleAvatar'
            url={currentUser && currentUser.avatarUrl}
          />
        </div>
        <div styleName='body-column'>
          <input
            type='text'
            styleName='titleInput'
            placeholder={titlePlaceholder}
            value={title || ''}
            onChange={this.handleTitleChange}
            disabled={loading}
            ref={x => { this.titleInput = x }}
            maxLength={MAX_TITLE_LENGTH}
          />
          {titleLengthError && <span styleName='title-error'>{`Title can't have more than ${MAX_TITLE_LENGTH} characters`}</span>}
          <HyloEditor
            styleName='editor'
            placeholder={detailsPlaceholder}
            onChange={this.handleDetailsChange}
            contentHTML={details}
            readOnly={loading}
            parentComponent={'PostEditor'}
            ref={component => { this.editor = component && component.getWrappedInstance() }}
          />
        </div>
      </div>
      {linkPreview &&
        <LinkPreview linkPreview={linkPreview} onClose={this.removeLinkPreview} />}
      <AttachmentManager postId={id || 'new'} type='image' />
      <AttachmentManager postId={id || 'new'} type='file' />
      <div styleName='footer'>
        <div styleName='footerSection'>
          <div styleName='footerSection-label'>Topics</div>
          <div styleName='footerSection-communities'>
            <TopicSelector
              selectedTopics={topics}
              detailsTopics={detailsTopics}
              ref={component => { this.topicSelector = component && component.getWrappedInstance() }} />
          </div>
        </div>
        <div styleName='footerSection'>
          <div styleName='footerSection-label'>Post in</div>
          <div styleName='footerSection-communities'>
            <CommunitiesSelector
              options={communityOptions}
              selected={communities}
              onChange={this.setSelectedCommunities}
              readOnly={loading}
              ref={component => { this.communitiesSelector = component }}
            />
          </div>
        </div>
        <ActionsBar
          id={id}
          addImage={addImage}
          showImages={showImages}
          addFile={addFile}
          showFiles={showFiles}
          valid={valid}
          loading={loading}
          submitButtonLabel={this.buttonLabel()}
          save={() => this.save()}
          setAnnouncement={setAnnouncement}
          announcementSelected={announcementSelected}
          canModerate={canModerate}
          toggleAnnouncementModal={this.toggleAnnouncementModal}
          showAnnouncementModal={showAnnouncementModal}
          communityCount={get('communities', post).length}
          myModeratedCommunities={myModeratedCommunities}
          communities={post.communities}
        />
      </div>
    </div>
  }
}

export function ActionsBar ({id,
  addImage,
  showImages,
  addFile,
  showFiles,
  valid,
  loading,
  submitButtonLabel,
  save,
  setAnnouncement,
  announcementSelected,
  canModerate,
  toggleAnnouncementModal,
  showAnnouncementModal,
  communityCount,
  myModeratedCommunities,
  communities
}) {
  return <div styleName='actionsBar'>
    <div styleName='actions'>
      <ChangeImageButton update={addImage}
        uploadSettings={uploadSettings(id)}
        attachmentType='image'
        disable={showImages}>
        <Icon name='AddImage'
          styleName={cx('action-icon', {'highlight-icon': showImages})} />
      </ChangeImageButton>
      <ChangeImageButton update={addFile}
        uploadSettings={uploadSettings(id)}
        attachmentType='file'
        disable={showFiles}>
        <Icon name='Paperclip'
          styleName={cx('action-icon', {'highlight-icon': showFiles})} />
      </ChangeImageButton>
      {canModerate && <span data-tip='Send Announcement' data-for='announcement-tt'>
        <Icon name='Announcement'
          onClick={() => setAnnouncement(!announcementSelected)}
          styleName={cx('action-icon', {'highlight-icon': announcementSelected})}
        />
        <ReactTooltip
          effect={'solid'}
          delayShow={550}
          id='announcement-tt' />
      </span>}
      {showAnnouncementModal && <SendAnnouncementModal
        closeModal={toggleAnnouncementModal}
        save={save}
        communityCount={communityCount}
        myModeratedCommunities={myModeratedCommunities}
        communities={communities}
      />}

    </div>
    <Button
      onClick={announcementSelected ? toggleAnnouncementModal : save}
      disabled={!valid || loading}
      styleName='postButton'
      label={submitButtonLabel}
      color='green'
    />
  </div>
}
