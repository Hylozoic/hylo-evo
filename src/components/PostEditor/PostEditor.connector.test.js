import { mapStateToProps, mapDispatchToProps, mergeProps } from './PostEditor.connector'
import orm from 'store/models'
import { CREATE_POST } from './PostEditor.store'

let state
beforeAll(() => {
  const session = orm.session(orm.getEmptyState())
  const community = session.Community.create({id: '99', slug: 'foo'})
  session.LinkPreview.create({
    id: 1
  })

  session.Me.create({
    id: '1',
    memberships: [session.Membership.create({
      id: '345',
      community: community.id,
      hasModeratorRole: true
    })]
  })

  state = {
    orm: session.state,
    PostEditor: {
      linkPreviewStatus: false
    },
    pending: {
      FETCH_LINK_PREVIEW: false
    },
    AttachmentManager: {
      image: [],
      file: []
    }
  }
})

describe('mapStateToProps', () => {
  it('returns the right keys for a new post', () => {
    const props = {
      forNew: true,
      match: {
        params: {
          slug: 'foo'
        }
      }
    }
    expect(mapStateToProps(state, props)).toMatchSnapshot()
  })
  it('returns the right keys for a new post while pending', () => {
    const props = {
      forNew: true,
      match: {
        params: {
          slug: 'foo'
        }
      }
    }

    const newState = {
      ...state,
      pending: {
        FETCH_LINK_PREVIEW: false,
        [CREATE_POST]: true
      }
    }
    expect(mapStateToProps(newState, props)).toMatchSnapshot()
  })
})

describe('mapDispatchToProps', () => {
  it('returns the right keys', () => {
    expect(mapDispatchToProps()).toMatchSnapshot()
  })
})

describe('mergeProps', () => {
  it('goToPost redirects to topic with topicName', () => {
    const stateProps = {
      topicName: 'thetopicname',
      slug: 'theslug'
    }
    const dispatchProps = {
      goToUrl: jest.fn()
    }
    const action = {
      payload: {
        data: {
          createPost: {
            id: 123
          }
        }
      }
    }
    const mergedProps = mergeProps(stateProps, dispatchProps)
    mergedProps.goToPost(action)
    expect(dispatchProps.goToUrl).toHaveBeenCalled()
    expect(dispatchProps.goToUrl.mock.calls).toMatchSnapshot()
  })

  it('goToPost redirects to community feed with no topicName', () => {
    const stateProps = {
      slug: 'theslug'
    }
    const dispatchProps = {
      goToUrl: jest.fn()
    }
    const action = {
      payload: {
        data: {
          createPost: {
            id: 123
          }
        }
      }
    }
    const mergedProps = mergeProps(stateProps, dispatchProps)
    mergedProps.goToPost(action)
    expect(dispatchProps.goToUrl).toHaveBeenCalled()
    expect(dispatchProps.goToUrl.mock.calls).toMatchSnapshot()
  })
})
