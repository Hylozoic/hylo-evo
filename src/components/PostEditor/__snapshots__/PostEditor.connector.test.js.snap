// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapStateToProps returns the right keys for a new post 1`] = `
Object {
  "communityOptions": Array [
    SessionBoundModel {
      "_fields": Object {
        "id": "99",
        "members": undefined,
        "moderators": undefined,
        "posts": undefined,
        "slug": "foo",
      },
      "slug": "foo",
    },
  ],
  "currentCommunity": SessionBoundModel {
    "_fields": Object {
      "id": "99",
      "members": undefined,
      "moderators": undefined,
      "posts": undefined,
      "slug": "foo",
    },
    "slug": "foo",
  },
  "currentUser": SessionBoundModel {
    "_fields": Object {
      "id": "1",
      "messageThreads": undefined,
      "notifications": undefined,
      "posts": undefined,
      "skills": undefined,
    },
    "id": "1",
  },
  "editing": false,
  "fetchLinkPreviewPending": undefined,
  "files": Array [],
  "images": Array [],
  "linkPreview": null,
  "linkPreviewStatus": false,
  "loading": false,
  "post": null,
  "postPending": false,
  "showFiles": false,
  "showImages": false,
}
`;

exports[`mapStateToProps returns the right keys for a new post while pending 1`] = `
Object {
  "communityOptions": Array [
    SessionBoundModel {
      "_fields": Object {
        "id": "99",
        "members": undefined,
        "moderators": undefined,
        "posts": undefined,
        "slug": "foo",
      },
      "slug": "foo",
    },
  ],
  "currentCommunity": SessionBoundModel {
    "_fields": Object {
      "id": "99",
      "members": undefined,
      "moderators": undefined,
      "posts": undefined,
      "slug": "foo",
    },
    "slug": "foo",
  },
  "currentUser": SessionBoundModel {
    "_fields": Object {
      "id": "1",
      "messageThreads": undefined,
      "notifications": undefined,
      "posts": undefined,
      "skills": undefined,
    },
    "id": "1",
  },
  "editing": true,
  "fetchLinkPreviewPending": undefined,
  "files": Array [],
  "images": Array [],
  "linkPreview": null,
  "linkPreviewStatus": false,
  "loading": true,
  "post": null,
  "postPending": true,
  "showFiles": false,
  "showImages": false,
}
`;
